components/builder-worker/src/vcs.rs

```diff
@@ -21,24 +21,90 @@ use error::Result;
 pub struct VCS {
     pub vcs_type: String,
     pub data: String,
+    pub auth_token: Option<String>,
+    pub username: Option<String>,
 }
 
 impl VCS {
-    pub fn new(vcs_type: String, data: String) -> VCS {
+    pub fn new(
+        vcs_type: String,
+        data: String,
+        auth_token: Option<String>,
+        username: Option<String>,
+    ) -> VCS {
         VCS {
             vcs_type: vcs_type,
             data: data,
+            auth_token: auth_token,
+            username: username,
         }
     }
 
     pub fn clone(&self, path: &Path) -> Result<()> {
         match self.vcs_type.as_ref() {
             "git" => {
-                debug!("cloning git repository, url={}, path={:?}", self.data, path);
-                git2::Repository::clone(&self.data, path)?;
+                debug!(
+                    "cloning git repository, url={}, path={:?}",
+                    self.url(),
+                    path
+                );
+                git2::Repository::clone(&self.url(), path)?;
                 Ok(())
             }
             _ => panic!("Unknown vcs type"),
         }
     }
+
+    pub fn url(&self) -> String {
+        // git urls come from the clone_url of a repo which should always
```

Comment:
I think we want to play it safe here and parse the data and manipulate it that way. I would use this [crate](https://github.com/servo/rust-url) and these functions:

* [parse](https://docs.rs/url/1.5.1/url/struct.Url.html#method.parse)
* [set_username](https://docs.rs/url/1.5.1/url/struct.Url.html#method.set_username)
* [set_password](https://docs.rs/url/1.5.1/url/struct.Url.html#method.set_password)

We should validate that the "Data" in the constructor of the vcs struct before we get to this function